File: typescript/src/transport.ts
// Byte transport for SanDBox's stdio protocol: moves bytes and owns
// connection lifetime. Knows nothing about JSON Lines framing semantics or
// the op vocabulary -- that's codec.ts's job. This layer is exactly where
// I/O, subprocess, and socket machinery are allowed to live
// (.claude/rules/architecture.md).

import { type ChildProcess, spawn } from "node:child_process";
import { createConnection, type Socket } from "node:net";
import { constants as osConstants } from "node:os";
import type { Duplex, Readable, Writable } from "node:stream";
import {
  MAX_LINE_BYTES,
  ProtocolError,
  SanDBoxError,
  SanDBoxTimeoutError,
} from "./codec.js";

export interface Transport {
  writeLine(line: Uint8Array): Promise<void>;
  readLine(timeoutMs: number | null, signal?: AbortSignal): Promise<Buffer>;
  close(timeoutMs: number): Promise<void>;
}

const LF = 0x0a;

/** One reader per connection, feeding readLine() callers a single
 *  line at a time. Backpressure is 1 slot deep: once a complete line is
 *  buffered the source stream is paused, and resumed only once that line
 *  is taken (the direct translation of Python's queue.Queue(maxsize=1)),
 *  so a misbehaving peer can't make this driver buffer unboundedly.
 *
 *  A terminal condition (EOF or a stream error) is memoized and re-thrown
 *  on every subsequent readLine -- once the stream ends there is no way to
 *  "get another line" from it, so every later call must see the same
 *  outcome rather than hang. */
class LineReader {
  #source: Readable;
  #pending: Buffer[] = [];
  #pendingBytes = 0;
  #queue: Buffer[] = [];
  #waiters: Array<{ resolve: (line: Buffer) => void; reject: (err: Error) => void }> =
    [];
  #terminal: Error | null = null;

  constructor(source: Readable) {
    this.#source = source;
    source.on("data", (chunk: Buffer) => this.#onData(chunk));
    source.on("end", () =>
      this.#terminate(new SanDBoxError("connection closed (EOF)")),
    );
    source.on("error", (err: Error) => this.#terminate(err));
  }

  #onData(chunk: Buffer): void {
    // Once terminal (EOF, an error, or a MAX_LINE_BYTES violation), ignore
    // any further data. Without this guard a peer that keeps writing after
    // a protocol violation -- or simply keeps writing forever without a
    // newline -- would make #pending grow without bound and keep this
    // callback running flat-out even though nothing can ever be delivered
    // again.
    if (this.#terminal) return;
    let start = 0;
    for (let i = 0; i < chunk.length; i++) {
      if (chunk[i] !== LF) continue;
      const piece = chunk.subarray(start, i);
      this.#pending.push(piece);
      const line = this.#pending.length === 1 ? piece : Buffer.concat(this.#pending);
      this.#pending = [];
      this.#pendingBytes = 0;
      this.#deliver(line);
      start = i + 1;
    }
    if (start < chunk.length) {
      const rest = chunk.subarray(start);
      this.#pending.push(rest);
      this.#pendingBytes += rest.length;
      if (this.#pendingBytes > MAX_LINE_BYTES) {
        this.#terminate(
          new ProtocolError(
            `response line exceeds ${MAX_LINE_BYTES} bytes without a newline`,
          ),
        );
        return;
      }
    }
    // 1-slot backpressure: stop reading more once a line is queued and
    // unclaimed.
    if (this.#queue.length > 0) {
      this.#source.pause();
    }
  }

  #deliver(line: Buffer): void {
    const waiter = this.#waiters.shift();
    if (waiter) {
      waiter.resolve(line);
    } else {
      this.#queue.push(line);
      this.#source.pause();
    }
  }

  #terminate(err: Error): void {
    if (this.#terminal) return;
    this.#terminal = err;
    // Nothing will ever be delivered again -- drop any partial line and
    // stop consuming the source so a peer that keeps writing (deliberately
    // or not) can't keep this reader spinning or growing #pending forever.
    // pause() also applies OS-level backpressure: a still-writing child
    // blocks on its next write() instead of burning CPU indefinitely.
    this.#pending = [];
    this.#pendingBytes = 0;
    this.#source.pause();
    const waiters = this.#waiters;
    this.#waiters = [];
    for (const w of waiters) w.reject(err);
  }

  /** Forces the terminal condition immediately, for a failure that doesn't
   *  arrive through the source stream itself -- e.g. the child process
   *  failing to spawn at all, in which case stdout will never emit
   *  anything on its own. */
  fail(err: Error): void {
    this.#terminate(err);
  }

  /** Resolves with the next complete line, or rejects with SanDBoxTimeoutError
   *  if timeoutMs elapses first (null = wait forever), or with the memoized
   *  terminal error once the stream has ended. */
  readLine(timeoutMs: number | null, signal?: AbortSignal): Promise<Buffer> {
    const queued = this.#queue.shift();
    if (queued !== undefined) {
      if (this.#queue.length === 0) this.#source.resume();
      return Promise.resolve(queued);
    }
    if (this.#terminal) return Promise.reject(this.#terminal);
    if (signal?.aborted) return Promise.reject(signal.reason ?? new Error("aborted"));

    return new Promise<Buffer>((resolve, reject) => {
      const waiter = { resolve, reject };
      this.#waiters.push(waiter);

      let timer: NodeJS.Timeout | undefined;
      const onAbort = () =>
        settle(() => reject(signal?.reason ?? new Error("aborted")));
      const cleanup = () => {
        if (timer) clearTimeout(timer);
        signal?.removeEventListener("abort", onAbort);
        const idx = this.#waiters.indexOf(waiter);
        if (idx !== -1) this.#waiters.splice(idx, 1);
      };
      const settle = (run: () => void) => {
        cleanup();
        run();
      };

      waiter.resolve = (line) => settle(() => resolve(line));
      waiter.reject = (err) => settle(() => reject(err));

      if (timeoutMs !== null) {
        timer = setTimeout(() => {
          settle(() =>
            reject(new SanDBoxTimeoutError("timed out waiting for a response line")),
          );
        }, timeoutMs);
      }
      if (signal) {
        signal.addEventListener("abort", onAbort, { once: true });
      }
    });
  }
}

/** Copies src to dst in the background to completion, so a child's stderr
 *  pipe never fills and blocks the child (protocol.md: stderr must always
 *  be drained). Errors from either end (the sink closing first, etc.) are
 *  swallowed -- draining is best-effort, not a contract the caller can
 *  observe failing. */
function drain(src: Readable, dst: Writable): void {
  src.on("error", () => {});
  dst.on("error", () => {});
  src.pipe(dst, { end: false });
}

function afterDeadline<T>(promise: Promise<T>, ms: number): Promise<T | "timeout"> {
  return Promise.race([
    promise,
    new Promise<"timeout">((resolve) => setTimeout(() => resolve("timeout"), ms)),
  ]);
}

export interface DirectOptions {
  readonly env?: NodeJS.ProcessEnv;
  readonly cwd?: string | URL;
  /** Where the child's stderr goes. Omit it and stderr is routed to the OS
   *  null device (no pipe to fill, nothing to drain). Supply a Writable and
   *  it is piped with {end:false}, so passing process.stderr is safe (this
   *  transport won't close it). */
  readonly stderr?: Writable;
}

/** Direct-connect transport: launches `command` with `args` as a child
 *  process and speaks the protocol over its stdin/stdout. The command to
 *  launch is entirely caller-supplied, which is what lets local,
 *  `ssh user@host san-db-ox --serve-stdio`, `docker run -i`, and
 *  `kubectl exec -i` all use this one code path -- there is deliberately no
 *  SSH- or Docker-specific constructor (.claude/rules/architecture.md). */
export class DirectTransport implements Transport {
  #child: ChildProcess;
  #reader: LineReader;
  // Resolves on the child's "close" event, which (unlike "exit") fires only
  // once every stdio stream has finished emitting -- waiting on this rather
  // than "exit" is what guarantees a provided stderr sink has received
  // everything before close() tears the streams down.
  #closed: Promise<void>;
  #closePromise: Promise<void> | null = null;

  private constructor(child: ChildProcess, reader: LineReader, closed: Promise<void>) {
    this.#child = child;
    this.#reader = reader;
    this.#closed = closed;
  }

  static start(
    command: string,
    args: readonly string[] = [],
    options: DirectOptions = {},
  ): DirectTransport {
    const child = spawn(command, args, {
      stdio: ["pipe", "pipe", options.stderr ? "pipe" : "ignore"],
      env: options.env,
      cwd: options.cwd,
    });
    // A write to a dead child raises EPIPE as an "error" event on stdin;
    // left unhandled that is an uncaught exception that kills the process.
    // Swallow it here -- the memoized reader terminal error (from the
    // matching stdout "end"/"error", or from the spawn-failure handler
    // below) is what surfaces to the next call.
    child.stdin?.on("error", () => {});
    // biome-ignore lint/style/noNonNullAssertion: stdio[1] is "pipe"
    const reader = new LineReader(child.stdout!);
    // Spawn failure (e.g. ENOENT) arrives as an async "error" event, not a
    // synchronous throw -- unlike Python's Popen, which raises
    // FileNotFoundError immediately. Feed it into the reader as the
    // terminal condition so the first readLine (reading hello) rejects
    // with it, which is what surfaces to callers of connect().
    child.on("error", (err) => {
      reader.fail(
        new SanDBoxError(`starting ${JSON.stringify(command)}: ${err.message}`, {
          cause: err,
        }),
      );
    });
    if (options.stderr && child.stderr) {
      drain(child.stderr, options.stderr);
    }
    const closed = new Promise<void>((resolve) => {
      child.once("close", () => resolve());
    });
    return new DirectTransport(child, reader, closed);
  }

  async writeLine(line: Uint8Array): Promise<void> {
    const stdin = this.#child.stdin;
    if (!stdin) throw new SanDBoxError("child process has no stdin");
    await new Promise<void>((resolve, reject) => {
      stdin.write(line, (err) =>
        err
          ? reject(new SanDBoxError(`writing request: ${err.message}`, { cause: err }))
          : resolve(),
      );
    });
  }

  readLine(timeoutMs: number | null, signal?: AbortSignal): Promise<Buffer> {
    return this.#reader.readLine(timeoutMs, signal);
  }

  /** The child's exit code. null while still running. Negative (-signal)
   *  when the child was terminated by a signal, mirroring Python's
   *  Popen.poll() convention (SIGTERM -> -15) -- Node reports exitCode and
   *  signalCode separately, and returning exitCode raw would make null
   *  ambiguous between "still running" and "killed by a signal". */
  get exitCode(): number | null {
    if (this.#child.exitCode !== null) return this.#child.exitCode;
    if (this.#child.signalCode) {
      const n =
        osConstants.signals[this.#child.signalCode as keyof typeof osConstants.signals];
      return typeof n === "number" ? -n : null;
    }
    return null;
  }

  /** Staged shutdown: close stdin, wait; SIGTERM, wait; SIGKILL, wait
   *  unconditionally (SIGKILL cannot be ignored). Safe to call more than
   *  once, and safe to call after the process already exited on its own
   *  (e.g. after a successful overwrite, which ends the connection from
   *  the server's side). */
  close(timeoutMs: number): Promise<void> {
    if (this.#closePromise) return this.#closePromise;
    this.#closePromise = this.#close(timeoutMs);
    return this.#closePromise;
  }

  async #close(timeoutMs: number): Promise<void> {
    const child = this.#child;
    if (child.exitCode === null && child.signalCode === null) {
      child.stdin?.end();
      if ((await afterDeadline(this.#closed, timeoutMs)) === "timeout") {
        child.kill("SIGTERM");
        if ((await afterDeadline(this.#closed, timeoutMs)) === "timeout") {
          child.kill("SIGKILL");
          await this.#closed;
        }
      }
    } else {
      // Already exited (e.g. after a successful overwrite), but "close"
      // (unlike "exit") may not have fired yet if stdio hasn't finished
      // flushing -- wait for it too, bounded by timeoutMs, so a provided
      // stderr sink still receives everything before the streams below are
      // torn down.
      await afterDeadline(this.#closed, timeoutMs);
    }
    child.stdout?.destroy();
    child.stderr?.destroy();
    child.stdin?.destroy();
  }
}

/** Races a freshly created socket's "connect" event against a deadline.
 *  Deliberately not socket.setTimeout(), which is an idle timeout applied
 *  after connecting, not a bound on the connection attempt itself. */
function connectWithDeadline(
  create: () => Socket,
  timeoutMs: number | null,
): Promise<Socket> {
  return new Promise((resolve, reject) => {
    const socket = create();
    let settled = false;
    let timer: NodeJS.Timeout | undefined;

    const settle = (run: () => void) => {
      if (settled) return;
      settled = true;
      socket.removeListener("connect", onConnect);
      socket.removeListener("error", onError);
      if (timer) clearTimeout(timer);
      run();
    };
    const onConnect = () => settle(() => resolve(socket));
    const onError = (err: Error) =>
      settle(() =>
        reject(new SanDBoxError(`connecting: ${err.message}`, { cause: err })),
      );

    socket.once("connect", onConnect);
    socket.once("error", onError);
    if (timeoutMs !== null) {
      timer = setTimeout(() => {
        settle(() => {
          socket.destroy();
          reject(new SanDBoxTimeoutError(`timed out connecting after ${timeoutMs}ms`));
        });
      }, timeoutMs);
    }
  });
}

/** Socket transport: TCP, UNIX domain socket, or any already-connected
 *  Duplex (the seam for connections this module can't build directly, most
 *  notably a TLS socket for mutual-TLS authentication -- no TLS-specific
 *  constructor exists here, and this module never imports "node:tls"). No
 *  staged shutdown: there is no child process on this end to signal or
 *  reap, so close is a single deadline-then-destroy. */
export class SocketTransport implements Transport {
  #stream: Duplex;
  #reader: LineReader;
  #closePromise: Promise<void> | null = null;

  constructor(stream: Duplex) {
    this.#stream = stream;
    this.#reader = new LineReader(stream);
  }

  static async connectTcp(
    host: string,
    port: number,
    timeoutMs: number | null,
  ): Promise<SocketTransport> {
    const socket = await connectWithDeadline(
      () => createConnection({ host, port }),
      timeoutMs,
    );
    return new SocketTransport(socket);
  }

  static async connectUnix(
    path: string,
    timeoutMs: number | null,
  ): Promise<SocketTransport> {
    const socket = await connectWithDeadline(
      () => createConnection({ path }),
      timeoutMs,
    );
    return new SocketTransport(socket);
  }

  async writeLine(line: Uint8Array): Promise<void> {
    await new Promise<void>((resolve, reject) => {
      this.#stream.write(line, (err) =>
        err
          ? reject(new SanDBoxError(`writing request: ${err.message}`, { cause: err }))
          : resolve(),
      );
    });
  }

  readLine(timeoutMs: number | null, signal?: AbortSignal): Promise<Buffer> {
    return this.#reader.readLine(timeoutMs, signal);
  }

  close(timeoutMs: number): Promise<void> {
    if (this.#closePromise) return this.#closePromise;
    this.#closePromise = this.#close(timeoutMs);
    return this.#closePromise;
  }

  async #close(timeoutMs: number): Promise<void> {
    const stream = this.#stream;
    const closed = new Promise<void>((resolve) => {
      if (stream.destroyed) {
        resolve();
        return;
      }
      stream.once("close", () => resolve());
    });
    stream.end();
    await afterDeadline(closed, timeoutMs);
    stream.destroy();
  }
}

File: python/src/san_db_ox/_transport.py
"""Direct (subprocess) and Socket transports for SanDBox's stdio protocol.

A transport only knows how to read/write bytes and manage the connection's
lifetime; it knows nothing about JSON Lines framing or the op vocabulary
(that's _codec.py's job -- .claude/rules/architecture.md).

The background line-reading thread lives here, not in the client/session
layer above: how to unblock a stuck read differs per transport (kill the
child process for Direct, shutdown() the socket for Socket), so each
transport is responsible for its own read cancellation.
"""

from __future__ import annotations

import contextlib
import os
import queue
import socket as socket_module
import subprocess
import threading
from collections.abc import Mapping, Sequence
from typing import IO, Protocol

from san_db_ox._codec import MAX_LINE_BYTES, ProtocolError


class Transport(Protocol):
    """What every connection to a SanDBox process can do, regardless of
    transport. Structural (no common base class needed)."""

    def write_line(self, line: bytes) -> None: ...
    def read_line(self, timeout: float | None) -> bytes: ...
    def close(self, timeout: float) -> None: ...


class _LineReader:
    """Background thread reading newline-delimited lines from a byte stream
    into a backpressured queue.

    Mirrors Go's session.readLoop: responses are framed one per line and
    always arrive in order, so a single reader thread feeding one slot is
    enough -- callers never race independent reads against each other.
    """

    def __init__(self, stream: IO[bytes]) -> None:
        self._stream = stream
        self._queue: queue.Queue[bytes | BaseException] = queue.Queue(maxsize=1)
        # queue.Queue has no close(): once the terminal item (EOF/error) is
        # put, a second get() would block forever unless we remember it
        # ourselves and keep handing it back.
        self._terminal: BaseException | None = None
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()

    def _run(self) -> None:
        try:
            while True:
                line = self._stream.readline(MAX_LINE_BYTES + 1)
                if not line:
                    self._queue.put(EOFError("connection closed"))
                    return
                if not line.endswith(b"\n"):
                    self._queue.put(
                        ProtocolError(
                            f"response line exceeds {MAX_LINE_BYTES} bytes "
                            "without a newline"
                        )
                    )
                    return
                self._queue.put(line[:-1])
        except OSError as e:
            self._queue.put(e)

    def read_line(self, timeout: float | None) -> bytes:
        if self._terminal is not None:
            raise self._terminal
        try:
            item = self._queue.get(timeout=timeout)
        except queue.Empty:
            raise TimeoutError("timed out waiting for a response line") from None
        if isinstance(item, BaseException):
            self._terminal = item
            raise item
        return item

    def join(self, timeout: float | None) -> None:
        self._thread.join(timeout=timeout)


def _drain(src: IO[bytes], dst: IO[bytes]) -> None:
    """Copy src to dst until EOF, swallowing errors from either side
    closing first. Used for stderr, which must always be drained
    (protocol.md) even when the caller doesn't want to see it."""
    try:
        while True:
            chunk = src.read(65536)
            if not chunk:
                return
            dst.write(chunk)
    except (OSError, ValueError):
        return


class DirectTransport:
    """Direct-connect transport: a child process, talked to over its
    stdin/stdout. Mirrors go/sandbox/internal/transport/direct.go.
    """

    def __init__(
        self, proc: subprocess.Popen[bytes], stderr_thread: threading.Thread | None
    ) -> None:
        assert proc.stdin is not None
        assert proc.stdout is not None
        self._proc = proc
        self._stdin: IO[bytes] = proc.stdin
        self._stdout: IO[bytes] = proc.stdout
        self._reader = _LineReader(proc.stdout)
        self._stderr_thread = stderr_thread

    @classmethod
    def start(
        cls,
        command: str,
        args: Sequence[str] = (),
        *,
        env: Mapping[str, str] | None = None,
        cwd: str | os.PathLike[str] | None = None,
        stderr: IO[bytes] | None = None,
    ) -> DirectTransport:
        # stderr is always drained -- even when discarded -- so a chatty
        # server can never block on a full pipe (protocol.md). DEVNULL has
        # no pipe to fill in the first place, so a background drain thread
        # is only needed when a caller actually wants the bytes.
        stderr_pipe = subprocess.PIPE if stderr is not None else subprocess.DEVNULL
        proc = subprocess.Popen(
            [command, *args],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=stderr_pipe,
            env=dict(env) if env is not None else None,
            cwd=cwd,
        )
        stderr_thread = None
        if stderr is not None:
            assert proc.stderr is not None
            stderr_thread = threading.Thread(
                target=_drain, args=(proc.stderr, stderr), daemon=True
            )
            stderr_thread.start()
        return cls(proc, stderr_thread)

    def write_line(self, line: bytes) -> None:
        # protocol.md: flush after every request. Popen's pipes are
        # buffered (unlike Go's raw os.Pipe), so this is required, not
        # optional.
        self._stdin.write(line)
        self._stdin.flush()

    def read_line(self, timeout: float | None) -> bytes:
        return self._reader.read_line(timeout)

    @property
    def exit_code(self) -> int | None:
        return self._proc.poll()

    def close(self, timeout: float) -> None:
        # Staged shutdown, matching direct.go: stdin close -> wait ->
        # SIGTERM -> wait -> SIGKILL. Popen.terminate()/.kill() already
        # abstract the POSIX/Windows signal difference that direct.go
        # handles by hand.
        with contextlib.suppress(OSError):
            self._stdin.close()
        try:
            self._proc.wait(timeout=timeout)
        except subprocess.TimeoutExpired:
            self._proc.terminate()
            try:
                self._proc.wait(timeout=timeout)
            except subprocess.TimeoutExpired:
                self._proc.kill()
                self._proc.wait()
        if self._stderr_thread is not None:
            self._stderr_thread.join(timeout=timeout)
            assert self._proc.stderr is not None
            with contextlib.suppress(OSError):
                self._proc.stderr.close()
        # The process has exited by now, so its end of the pipe is closed
        # and the reader thread's blocking readline() has already returned
        # (or is about to) -- join before closing the stream out from under
        # it.
        self._reader.join(timeout=timeout)
        with contextlib.suppress(OSError):
            self._stdout.close()


class SocketTransport:
    """Socket transport: a TCP/UNIX socket, typically fronted by something
    like socat. Mirrors go/sandbox/internal/transport/socket.go.

    Unlike DirectTransport, close has no staged escalation: there is no
    child process to signal or reap here.
    """

    def __init__(self, sock: socket_module.socket) -> None:
        self._sock = sock
        self._file: IO[bytes] = sock.makefile("rb")
        self._reader = _LineReader(self._file)

    @classmethod
    def connect_tcp(
        cls, host: str, port: int, *, timeout: float | None
    ) -> SocketTransport:
        sock = socket_module.create_connection((host, port), timeout=timeout)
        sock.settimeout(None)  # back to blocking mode for the reader thread
        return cls(sock)

    @classmethod
    def connect_unix(
        cls, path: str | os.PathLike[str], *, timeout: float | None
    ) -> SocketTransport:
        sock = socket_module.socket(socket_module.AF_UNIX, socket_module.SOCK_STREAM)
        sock.settimeout(timeout)
        sock.connect(os.fspath(path))
        sock.settimeout(None)
        return cls(sock)

    def write_line(self, line: bytes) -> None:
        self._sock.sendall(line)

    def read_line(self, timeout: float | None) -> bytes:
        return self._reader.read_line(timeout)

    def close(self, timeout: float) -> None:
        try:
            self._sock.settimeout(timeout)
            # shutdown() (unlike close()) reliably unblocks a concurrent
            # blocking recv() in the reader thread on every platform we
            # support -- Go achieves the same with SetDeadline.
            self._sock.shutdown(socket_module.SHUT_RDWR)
        except OSError:
            pass
        self._reader.join(timeout=timeout)
        with contextlib.suppress(OSError):
            self._file.close()
        self._sock.close()
